model_hub/examples/huggingface/token-classification/ner_trial.py

```diff
@@ -0,0 +1,118 @@
+        self.hparams.num_labels = datasets_metadata.num_labels
+
+        # Parse hparams and init model, optimizer, and lr_scheduler.
+        # See model_hub/huggingface/_arg_parser.py for default expected args.
+        config_args, tokenizer_args, model_args = hf.default_parse_config_tokenizer_model_args(
+            self.hparams
+        )
+        optimizer_args, scheduler_args = hf.default_parse_optimizer_lr_scheduler_args(self.hparams)
+        self.logger.info("Config arg:")
+        self.logger.info(config_args)
+        self.logger.info("Tokenizer arg:")
+        self.logger.info(tokenizer_args)
+        self.logger.info("Model arg:")
+        self.logger.info(model_args)
+        self.logger.info("Optimizer arg:")
+        self.logger.info(optimizer_args)
+        self.logger.info("LR Scheduler arg:")
+        self.logger.info(scheduler_args)
+
+        # This is used in the backward step called in train_batch of BaseTransformerTrial.
+        self.grad_clip_fn = (
+            lambda x: torch.nn.utils.clip_grad_norm_(x, optimizer_args.max_grad_norm)
+            if optimizer_args.max_grad_norm > 0
+            else None
+        )
+
+        self.config, self.tokenizer, self.model = hf.build_using_auto(
+            config_args, tokenizer_args, self.hparams.model_mode, model_args
+        )
+        self.model = self.context.wrap_model(self.model)
+
+        # The map function for datasets requires all objects be pickle-able for now.  This means
+        # we need to call this function before we create the optimizer, which is not pickle-able.
+        # Once this requirement is removed from huggingface datasets, we can simply call
+        # super(NERTrial, self).__init__(context) before creating the tokenized datasets.
+        # See https://github.com/huggingface/datasets/pull/1703.
+        self.tokenized_datasets = ner_utils.build_tokenized_datasets(
+            self.raw_datasets,
+            self.model,
+            self.data_config,
+            self.tokenizer,
+            datasets_metadata.text_column_name,
+            datasets_metadata.label_column_name,
+            datasets_metadata.label_to_id,
+        )
+
+        self.optimizer = self.context.wrap_optimizer(
+            hf.build_default_optimizer(self.model, optimizer_args)
+        )
+
+        if self.hparams.use_apex_amp:
+            self.model, self.optimizer = self.context.configure_apex_amp(
+                models=self.model,
+                optimizers=self.optimizer,
+            )
+
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
+            hf.build_default_lr_scheduler(self.optimizer, scheduler_args),
+            LRScheduler.StepMode.STEP_EVERY_BATCH,
+        )
+
+        # Create metric reducer
+        self.reducer = context.experimental.wrap_reducer(
+            utils.PredLabelFnReducer(
+                functools.partial(ner_utils.compute_metrics, datasets_metadata.label_list)
+            ),
```

Comment:
Seems like neither PredLabelFnReducer nor compute_metrics are really useful to the user until they are combined.  Why not just offer a Reducer class that takes a `label_list` and does the compute_metrics entirely internally?

Reply:
PredLabelFnReducer is fairly general and works with any function for metric computation that takes as input predictions and labels.  This reducer is actually used in other examples with different compute_metrics functions that have cleaner interfaces (i.e. no need to wrap with partial filled in label_list).

Reply:
removed this MetricReducer.  agree that the simplereducer route is sufficient here.